### PR TITLE
Don't add phobos library dependences with -defaultlib=

### PR DESCRIPTION
**Pull request by marler8997**
This PR depends on:
https://github.com/dlang/phobos/pull/7021 [MERGED]
https://github.com/dlang/druntime/pull/2619 [MERGED]
https://github.com/dlang/phobos/pull/7050 [MERGED]
https://github.com/dlang/druntime/pull/2625 [MERGED]

### CODE REVIEW

**Comment by rainers** [comment]
Maybe `-defaultlib=`and  `-debuglib=` should be modified to also not pass the system libraries to the linker? I guess they are added only to support linking in what might be used by phobos.

**Comment by JinShil** [comment]
What about moving the default libs to `dmd.conf`:  https://github.com/dlang/installer/blob/master/create_dmd_release/extras/linux/dmd2/linux/bin64/dmd.conf

Then DMD doesn't need to hard-code all of the default libraries and it can be customized on a per-distribution/platform basis.  If users don't want to link in any default libraries, they just use `dmd -conf=` to remove the effect of the `dmd.conf` file.

**Comment by marler8997** [comment]
I have use cases where I want to remove the default libraries without having to remove anything else that is currently in dmd.conf. I still want the default library paths provided in dmd.conf. I may still add the phobos2 library, probably via a pragmas lib.

So whether or not the libraries we're hard-coded, or in dmd.conf, I would still need this flag, making that change orthogonal.

**Comment by marler8997** [comment]
@rainers I didn't know about `-defaultlib=`.  Going to modify my PR per your suggestion.

**Comment by thewilsonator** [comment]
Still need a changelog entry anywhichway.

File: changelog/no_default_lib.dd [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Phobos/DRuntime library dependencies are no longer added when -defaultlib= is given
+
+Phobos/DRuntime library dependencies (i.e. "-lrt", "-ldl", "kernel32") will no longer be added to the linker command when those libraries are disabled with "-defaultlib=".
```
Comment (CyberShadow) [inline changelog/no_default_lib.dd]: Could you please expand this a bit to make it clear that this is a breaking change, which use cases it affects, and what action is required for those cases.

Comment (marler8997) [reply changelog/no_default_lib.dd]: Done

File: src/dmd/link.d [hunk 2]
```diff
@@ -152,6 +152,15 @@ version (Posix)
...
 }
 
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// TODO: I'm sure this function already exists somewhere???
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+bool isNullOrEmpty(const(char)* str) { return str is null || str[0] == '\0'; }
```
Comment (thewilsonator) [inline src/dmd/link.d]: To avoid this you could always make global.params.symdebug`, `global.params.debuglibname` and `global.params.defaultlibname` `const(char)[]`s :) (in a separate PR of course)

Comment (marler8997) [reply src/dmd/link.d]: yeah, I guess for now I'll just detect when it's empty and normalize it to null

**Comment by marler8997** [comment]
Closing and re-opening to trigger autotester retry

File: test/compilable/nodefaultlibs.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#
+# Compile with/without -defaultlib=, make sure that the libraries
+# that were linked without -defaultlib= were not linked with it
+#
+
+mkdir -p ${OUTPUT_BASE}
+$DMD -m${MODEL} -of${OUTPUT_BASE}a -v ${EXTRA_FILES}/noruntime.d              > ${OUTPUT_BASE}/verbose_out_a
+$DMD -m${MODEL} -of${OUTPUT_BASE}b -v ${EXTRA_FILES}/noruntime.d -defaultlib= > ${OUTPUT_BASE}/verbose_out_b
+
+set +x
```
Comment (wilzbach) [inline test/compilable/nodefaultlibs.sh]: This is already set by default by the Bash runner ;-)

File: test/compilable/nodefaultlibs.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    fi
+done
+
+echo "found these options in verbose output:$found"
+
+fail=false
+for lib in $found; do
+    if grep "\\$lib" ${OUTPUT_BASE}/verbose_out_b; then
+        echo "Error: found library \"$lib\" even with -defaultlib="
+        fail=true
```
Comment (wilzbach) [inline test/compilable/nodefaultlibs.sh]: Why not fail directly here?

File: changelog/no_default_lib.dd [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+This will break any project that satisfies these 3 criteria:
```
Comment (wilzbach) [inline changelog/no_default_lib.dd]:
```suggestion
This will break any project that satisfies these three criteria:
```

File: changelog/no_default_lib.dd [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+1. If the project depends on -lpthread, -lm, -lrt or -ldl but does not explicitly link to them
```
Comment (wilzbach) [inline changelog/no_default_lib.dd]:
```suggestion
1. If the project depends on `-lpthread`, `-lm`, `-lrt` or `-ldl` but does not explicitly link to them
```

File: changelog/no_default_lib.dd [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+2. If the project uses "-defaultlib="
```
Comment (wilzbach) [inline changelog/no_default_lib.dd]:
```suggestion
2. If the project uses `-defaultlib=`
```

File: changelog/no_default_lib.dd [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+3. If the project uses DMD to link
+
+A project fitting these 3 criteria is implicitly relying on the hard-coded libraries that DMD adds rather than declaring those dependencies explicitly.  To fix a project that fits these criteria, it should add any libraries that it depends on with `-L-l<library>`.
```
Comment (wilzbach) [inline changelog/no_default_lib.dd]:
```suggestion
A project fitting these three criteria is implicitly relying on the hard-coded libraries that DMD adds rather than declaring those dependencies explicitly.  To fix a project that fits these criteria, it should add any libraries that it depends on with `-L-l<library>`.
```

File: test/compilable/nodefaultlibs.sh (hunk 4 above)
Comment (marler8997) [reply test/compilable/nodefaultlibs.sh]: So you can see all the errors.

File: test/compilable/nodefaultlibs.sh [hunk 9]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
```
Comment (JinShil) [inline test/compilable/nodefaultlibs.sh]: I know this may be an unpopular opinion, but please use D instead of shell scripts for these kinds of tests so they are platform independent.

Comment (marler8997) [reply test/compilable/nodefaultlibs.sh]: Would love to, can you explain how I can write this test in D?

Comment (JinShil) [reply test/compilable/nodefaultlibs.sh]: It would require importing Phobos, using the facilities in `std.process` to invoke the compiler, and then examining the output.  The test itself could be invoked with the host's `rdmd`.  Importing Phobos into the test suite may be controversial.  Others would have to speak up about that.  I don't have any problem with it, and would even encourage it.
The basic premise is that users are having a hell of a time building DMD, druntime, and phobos on Windows, and running the test suite.  Part of the problem is Windows users have to install a Posix-like environment to make it work and it's not even a natural fit for Windows.  D (through Phobos) abstracts most of the platform dependencies, so the same code works on all platforms that D supports.  If the tests and build scripts are written in D, it should work everywhere D is supported.

Comment (marler8997) [reply test/compilable/nodefaultlibs.sh]: Yeah I've brought up the issues with using bash on windows quite a bit and have received little to no response. I can't even get anyone to publish instructions on how to run the test suite on windows.

But for now it's what we have. Maybe I'll come up with an idea the leadership will respond to...but will have to be another PR.

Comment (JinShil) [reply test/compilable/nodefaultlibs.sh]: [build.d](https://github.com/dlang/dmd/blob/master/src/build.d) works today on Linux, Windows, and MacOS.  However, we need to migrate more of the makefile targets to it, and then we need to do the same for druntime and Phobos.  If that interests you and are willing to volunteer your skills to move it forward, it would be most welcome.  Precedent is already there, so it should be less of a hassle now.

Comment (marler8997) [reply test/compilable/nodefaultlibs.sh]: Here's one idea: https://github.com/dlang/dmd/pull/9848

File: src/dmd/link.d [hunk 10]
```diff
@@ -289,8 +291,11 @@ public int runLINK()
         else
         {
             OutBuffer cmdbuf;
-            global.params.libfiles.push("user32");
-            global.params.libfiles.push("kernel32");
+            if (!phobosLibname)
```
Comment (rainers) [inline src/dmd/link.d]: I guess this should be the opposite.

File: samples/all.sh [hunk 11]
```diff
@@ -41,7 +41,7 @@ $(DMD) winsamp gdi32.lib winsamp.def
 winsamp
 
 # COM client/server example
-$(DMD) dserver.d chello.d dserver.def advapi32.lib ole32.lib user32.lib
+$(DMD) dserver.d chello.d dserver.def advapi32.lib ole32.lib user32.lib kernel32.lib
```
Comment (wilzbach) [inline samples/all.sh]: Why is this change necessary?

File: samples/win32.mak [hunk 12]
```diff
@@ -45,18 +45,18 @@ wc2:
 	.\wc2.exe wc2.d
 
 winsamp:
-	$(DMD) winsamp $(DFLAGS) gdi32.lib user32.lib winsamp.def
+	$(DMD) winsamp $(DFLAGS) gdi32.lib user32.lib kernel32.lib winsamp.def
```
Comment (wilzbach) [inline samples/win32.mak]: Shouldn't be necessary

**Review by wilzbach — COMMENTED** [review]
The tests shouldn't require an update as they don't use -defaultlib=

File: samples/win32.mak (hunk 12 above)
Comment (marler8997) [reply samples/win32.mak]: You're right, but I added them cause I was getting a failure so I assumed `-defaultlib=` was getting set somewhere...maybe in the `DFLAGS` environment variable or something.  I'll remove these changes and see what happens.

Comment (marler8997) [reply samples/win32.mak]: I had the logic swapped, it was only adding kernel32 if `-defaultlib=`. Fixed.

File: changelog/no_default_lib.dd (hunk 8 above)
Comment (CyberShadow) [reply changelog/no_default_lib.dd]: @thewilsonator Why did you resolve this? The suggestion looks valid and has not been applied.

Comment (marler8997) [reply changelog/no_default_lib.dd]: Just so there's no confusion, I just fixed this after @CyberShadow just noted that I had not made the change yet.

**Comment by marler8997** [comment]
I'm stumped on this one.  Looks like on the linux platforms where the host/target are different bit-widths (`Linux_32_64` and `Linux_64_32`), the linker is not finding the pthread library symbols.

https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=3663202&isPull=true
```
/home/braddr/sandbox/at-client/pull-3663202-Linux_32_64/dmd/generated/linux/release/32/dmd -m64 -fPIC -w -I../../src -I../../import -Isrc -defaultlib= -debuglib= -dip1000 -L-lpthread -L-lm -L/home/braddr/sandbox/at-client/pull-3663202-Linux_32_64/druntime/generated/linux/debug/64/libdruntime.a -g -debug -unittest -ofgenerated/linux/debug/64/unittest_assert src/unittest_assert.d
/home/braddr/sandbox/at-client/pull-3663202-Linux_32_64/druntime/generated/linux/debug/64/libdruntime.a(monitor__941_8c9.o): In function `_d_monitor_staticctor':
/home/braddr/sandbox/at-client/pull-3663202-Linux_32_64/druntime/src/rt/monitor_.d:158: undefined reference to `pthread_mutexattr_init'
...
```

Even though you can see we are explicitly providing `-L-lpthread` to DMD.  Maybe there's a way to select different pthread libraries for your targest host?  If anyone is familiar with this issue let me know.

**Comment by CyberShadow** [comment]
@wilzbach's suggestions for the changelog are still not applied and have been incorrectly marked as resolved by @thewilsonator.

**Comment by rainers** [comment]
> Even though you can see we are explicitly providing `-L-lpthread` to DMD. Maybe there's a way to select different pthread libraries for your targest host? If anyone is familiar with this issue let me know.

`ld` is known to be picky about the order of libraries passed to it on the command line, i.e. `-lpthread` should better be after the libraries that reference symbols from `libpthread.a`. Can you add them to the end of the command line that invokes the linker?

**Comment by rainers** [comment]
From the top of the "build dmd" task: the failing Linux runs use `GNU ld (GNU Binutils for Ubuntu) 2.26.1`, while the successful ones use `GNU ld version 2.25.1-31.base.66.amzn1`. No idea what the difference is or whether it's just a coincidence.

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
If you could convert the shell test to D, that would be great, the less of them we have the better.

**Comment by marler8997** [comment]
Yeah I'll convert the test to D.

**Comment by marler8997** [comment]
Ok bash test is now in D.  Still need to figure out the linker issue when testing druntime though.

**Comment by ZombineDev** [comment]
Yep, it looks like you need to do more adjustments to the Makefiles in `druntime/test/`, but apart from that it looks promising.

**Comment by marler8997** [comment]
I've added a commit to put the core libs (pthread/m/dl/rt) at the end of the link command, but still getting the undefined reference errors in the linker...

**Comment by marler8997** [comment]
Cleaning up old PRs that aren't going anywhere.
